server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -267,6 +269,33 @@ public void createSnapshot(final CreateSnapshotRequest request, final ActionList
         submitCreateSnapshotRequest(request, listener, repository, new Snapshot(repositoryName, snapshotId), repository.getMetadata());
     }
 
+    public void globalStateMetaData(final SnapshotGlobalStateRequest request, final ActionListener<SnapshotGlobalStateResponse> listener) {
+        String repositoryName = request.repository();
+        Repository repository = repositoriesService.repository(repositoryName);
+
+        final ListenableFuture<RepositoryData> repositoryDataListener = new ListenableFuture<>();
+        repository.getRepositoryData(repositoryDataListener);
+
+        final List<SnapshotId> snapshotIdsToLoad = new ArrayList<>();
+        repositoryDataListener.addListener(listener.delegateFailureAndWrap((delegate, repositoryData) -> {
+
+            final String snapshotName = request.snapshot();
+            final Optional<SnapshotId> matchingSnapshotId = repositoryData.getSnapshotIds()
+                .stream()
+                .filter(s -> snapshotName.equals(s.getName()))
+                .findFirst();
+            if (matchingSnapshotId.isPresent() == false) {
+                throw new SnapshotRestoreException(repositoryName, snapshotName, "snapshot does not exist");
+            }
+
+            snapshotIdsToLoad.add(matchingSnapshotId.get());
+        }));
+
+        Metadata globalMetadata = repository.getSnapshotGlobalMetadata(snapshotIdsToLoad.get(0));
+
+        listener.onResponse(new SnapshotGlobalStateResponse(globalMetadata));
+    }
```

Comment:
This is close but it's not quite going to work, you are relying on `repositoryDataListener` completing immediately but that won't be the case always (e.g. if the repository is read-only). Also it'd be a little better to keep the request/response types isolated to the network layers. Instead I think you need to do something more like this:

```suggestion
        public void getSnapshotGlobalMetadata(ActionListener<Metadata> listener, String repositoryName, String snapshotName) {
        final Repository repository = repositoriesService.repository(repositoryName);
        repository.getRepositoryData(
            listener.delegateFailureAndWrap(
                (delegate, repositoryData) -> threadPool.executor(ThreadPool.Names.SNAPSHOT_META)
                    .execute(ActionRunnable.supply(delegate, () -> {
                        for (final SnapshotId snapshotId : repositoryData.getSnapshotIds()) {
                            if (Objects.equals(snapshotId.getName(), snapshotName)) {
                                return repository.getSnapshotGlobalMetadata(snapshotId);
                            }
                        }
                        throw new ResourceNotFoundException("snapshot [%s] in repository [%s] not found", snapshotName, repositoryName);
                    }))
            )
        );
    }
```

Reply:
Separated the service logic implementation and transport layer. The response is coming as expected. Yes, you are correct I should have taken care as `repositoryDataListener` is of type `ListenableFuture`. 

It's quite interesting and intriguing to follow the flow of listeners. At the same time sometimes I mess up. I want to learn and understand more about these listener flows. Any suggestion or sources from which I can read and understand things from scratch??

Reply:
> I want to learn and understand more about these listener flows. Any suggestion or sources from which I can read and understand things from scratch??

There's no magic involved or anything, it's all just implemented as plain old Java code in the ES codebase. I recommend trying to understand all the methods on `org.elasticsearch.action.ActionListener` (and maybe `org.elasticsearch.action.ActionRunnable` too) - at least look at the Javadocs, their test suites, and how they're used throughout the rest of the system. There's some subtle differences in terms of exception handling but the basic idea is not especially deep.

Reply:
Yepp for sure! I will go through it in the manner you described. Any other changes in the above PR? Will move ahead accordingly. 

Meanwhile, if you can ping me any other issue that I can start off, it would be great.

Reply:
> Any other changes in the above PR?

I'll give some more detailed suggestions next week, but they will be fairly superficial since your PR is fundamentally doing the right thing. For now you should just add suitable tests.

Reply:
> For now you should just add suitable tests.

Oh, and docs too.

Reply:
Sure, will push test cases and docs as well within this week itself.